packages/xsnap/src/xsnap.js

```diff
@@ -96,7 +102,65 @@ export async function xsnap(options) {
   }
 
   /** @type {(opts: import('tmp').TmpNameOptions) => Promise<string>} */
-  const ptmpName = promisify(fs.tmpName);
+  const ptmpName = fs.tmpName && promisify(fs.tmpName);
+
+  const makeLoadSnapshotHandlerWithFS = async () => {
+    assert(snapshotStream);
+    const snapPath = await ptmpName({
+      template: `load-snapshot-${safeHintFromDescription(
+        snapshotDescription,
+      )}-XXXXXX.xss`,
+    });
+
+    const cleanup = async () => fs.unlink(snapPath);
+    const afterSpawn = async () => {};
+
+    try {
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      const tmpSnap = await fs.open(snapPath, 'w');
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await tmpSnap.writeFile(
+        // @ts-expect-error incorrect typings, does support AsyncIterable
+        snapshotStream,
+      );
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await tmpSnap.close();
+    } catch (e) {
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await cleanup();
+      throw e;
+    }
+
+    return harden({
+      snapPath,
+      afterSpawn,
+      cleanup,
+    });
+  };
+
+  const makeLoadSnapshotHandlerWithPipe = async () => {
+    let done = Promise.resolve();
+
+    const cleanup = async () => done;
+
+    /** @param {Writable} loadSnapshotsStream */
+    const afterSpawn = async loadSnapshotsStream => {
+      assert(snapshotStream);
+      const destStream = loadSnapshotsStream;
+
+      const sourceStream = Readable.from(snapshotStream);
+      sourceStream.pipe(destStream, { end: false });
+
+      done = finished(sourceStream);
+      done.catch(noop).then(() => sourceStream.unpipe(destStream));
+    };
+
+    return harden({
+      snapPath: `@8:${safeHintFromDescription(snapshotDescription)}`,
```

Comment:
This works (I'm guessing the related xsnap change was to limit the `atoi` to the separating colon), although I'm not sure we really need the snapshot ID on the command line (they're long and not very useful).

If we really need it to be available, we could stick it in an environment variable for the child process, where you could read it from `/proc/PID/environ`.

But, this is fine for now, no need to change it today.

Reply:
so for `atoi` while `C` seem to say behavior is undefined when reaching a non numeric char, POSIX defines that the behavior is the one of `strtol`, and stops at non-numeric chars. In a perfect world I'd go an be explicit about this on the xsnap-worker side, but I really don't want another PR just for that.

Also the description will only include the hash when loading the vat the first time. On snapshot reload it will only contain `vatID-snapPos`. The reason I wanted to include is because for forced snapshot reload, I can end up with 2 process with the exact same args during a reload, not knowing which is which is which. I also have manually tested the old worker retirement, but being able to spot a problem with different cmdline is helpful if more than one process remains.